### DISCUSSION TITLE
Constructor Redirect

### DISCUSSION THREAD

**Discussion by JKamsker** [dotnet/roslyn#69574] [Ideas]
Firstly, I'd like to apologize if this isn't the appropriate forum for this kind of feature proposal. If there's a more suitable venue for this discussion, please let me know.

Proposal:
    Constructor redirect

Summary:
    Allow for constructor arguments to be seemingly mirrored by a generic factory method.

Sample:
   ```csharp
// Notice here, that the context isn't required or even shown, its hidden
myfactory.Create<CommandType>("param1", "param2", "param3");

// or during dependency injection
services.AddTransient(sp => sp.GetRequiredService<MyFactory>().Create<CommandType>("param1", "param2", "param3"));
// or simpler (through extension methods). Notice that the DbContext is not required here either
services.AddTransient<CommandType>("param1", "param2", "param3");


class CommandType
{
    
    public CommandType(string param1, string param2, string param3, [FromServices]DbContext context)
    {
        // do something
    }
}

class MyFactory
{
    private readonly IServiceProvider _provider;
    public MyFactory(IServiceProvider provider)
    {
        _provider = provider;    
    }

    [RelayArguments]
    public T Create<T>(params object[] args)
    {
        return ActivatorUtilities.CreateInstance<T>(_provider, args);
    }
}
```

Alternatives:
- Source Generators:
    a source generator could be written to archive this goal, but source generators do not fully support method signature changes. (They do support it but visual studio does not update the generated code when the signature changes)
- Generic Factory:
    a generic factory could be written (like the one above, but without attributes) but this would require the developer to know the constructor parameters of the type they are trying to create.

- Manually Implementing the factory for each type:
    this is the current solution, but it requires a lot of boilerplate code to be written for each type that needs to be created.
    

Motivation:
    Having a factory method that can create instances of a type is already possible.
    Now, with intellisense support, it would be possible to create instances of a type without having to know the constructor parameters.

**Comment by KennethHoff** [comment]
This should be possible with a Roslyn Analyzer. I think it'd be relatively straightforward too. 

Find the generic type used, find the ctors of that type, and warn if there's a mismatch.

**Comment by JKamsker** [comment]
@KennethHoff I cannot imagine how it should work for intellisense suggestions which would be almost the whole point of such a feature.

**Reply by KennethHoff** [reply]
**Edit:** Apparently, there's actually something called a [`CompletionService`](https://github.com/dotnet/roslyn/blob/5f432fcbcf2373fccb988818719ac64be26cf019/src/Features/Core/Portable/Completion/CompletionService.cs#L30) that you could use. This is news to me :shrug: 

---

It wouldn't, no. The best you could do is create a Code Fixer for each constructor and then choose from therefore.

Let's say you have a `Person` class with the following ctors:

```c#
public class Person
{
    public Person(string firstName) {...}
    public Person(string firstName, string lastName) {...}
    public person(string firstName, string middleName, string lastName) {...}
}
```

You could add the following fixers:
 - Add `Person(string firstName)`
 - Add `Person(string firstName, string lastName)`
 - Add `Person(string firstName, string middleName, string lastName)`

I can almost guarantee that this feature will never be a part of C# "proper", so this is the best option you can have.

**Reply by JKamsker** [reply]
Thank you! I will certainly investigate this further. It may be feasible in the future, when the SourceGen support in Visual Studio isn't as poor and the produced code loads during runtime instead of just at start-up.

**Reply by jasonmalinowski** [reply]
Although I'm not sure I'm understanding the proposal, I'm confused by this statement:

> a source generator could be written to archive this goal, but source generators do not fully support method signature changes. (They do support it but visual studio does not update the generated code when the signature changes)

They absolutely can be reran if a signature changes...

**Reply by JKamsker** [reply]
https://github.com/JKamsker/ObjectFactorySourceGen/blob/master/tests/ObjectFactorySourceGen.ConsoleTest/CommandType.cs

This shows the Object that should be created by the factory. 
``    public CommandType(string myParameter, string myParameter1, string myParameter2, [FromServices] MyService1 context0, [FromServices] MyService context, [FromServices] MyService context1)``

The factory (generated to https://github.com/JKamsker/ObjectFactorySourceGen/blob/master/tests/ObjectFactorySourceGen.ConsoleTest/Program.cs#L44) only exposes the parameters that are not from services. 
``public CreateCommandType(string myParameter, string myParameter1, string myParameter2)``


The problem in my tests was, that new constructor overloads were not added to the factory while vs was running. They only appeared after restarting vs. 
After doing some research, it really seems like it is a vs problem which does not reload signatures of generated code.

It could also be that my source generator was not running fine but several posts on the internet suggest that i am not the only one with this problem. I would be more than happy if you or someone else could show me a way to fix this problem though.

**Reply by CyrusNajmabadi** [reply]
> After doing some research, it really seems like it is a vs problem which does not reload signatures of generated code.

We're unaware of any such problem.  Do you have a repro that shows this?  Tons of generators are running in VS today and successfully regenerating signatures all the time.  So something very weird must be going on with your project.

Can you link us and give a precise set of steps of what you did where the signatures did not update?

**Reply by JKamsker** [reply]
Sure! Ill get it ready asap! @CyrusNajmabadi

**Reply by JKamsker** [reply]
Well ok, turns out it was my fault all along. Thought it was an error in vs while it really was a bug hiding in my code, im sorry 😅
